src/lender/tranche/v3/tranche.sol

```diff
@@ -0,0 +1,169 @@
+    return currency.balanceOf(self);
+}
+
+function tokenSupply() external view returns (uint) {
+    return token.totalSupply();
+}
+
+function depend(bytes32 contractName, address addr) public auth {
+    if (contractName == "token") { token = ERC20Like(addr); }
+    if (contractName == "currency") { currency = ERC20Like(addr); }
+    if (contractName == "ticker") { ticker = TickerLike(addr); }
+    else revert();
+}
+
+// supplyOrder function can be used to place or revoke an supply 
+function supplyOrder(uint epochID, uint supplyAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    uint currentSupplyAmount = epochs[epochID].supplyCurrencyAmount[msg.sender];
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = supplyAmount;
+    epochs[epochID].totalSupply = safeAdd(safeSub(epochs[epochID].totalSupply, currentSupplyAmount), supplyAmount);
+    if (supplyAmount > currentSupplyAmount) {
+        uint delta = safeSub(supplyAmount, currentSupplyAmount);
+        require(currency.transferFrom(msg.sender, self, delta), "currency-transfer-failed");
+        return;
+    } 
+    uint delta = safeSub(currentSupplyAmount, supplyAmount);
+    if (delta > 0) {
+       require(currency.transferFrom(self, msg.sender, delta), "currency-transfer-failed");
+    }
+}
+
+// redeemOrder function can be used to place or revoke a redeem
+function redeemOrder(uint epochID, uint redeemAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    
+    uint currentRedeemAmount = epochs[epochID].redeemTokenAmount[msg.sender];
+    epochs[epochID].redeemTokenAmount[msg.sender] = redeemAmount;
+    epochs[epochID].totalRedeem = safeAdd(safeSub(epochs[epochID].totalRedeem, currentRedeemAmount), redeemAmount);
+    
+     if (redeemAmount > currentRedeemAmount) {
+        uint delta = safeSub(redeemAmount, currentRedeemAmount);
+        require(token.transferFrom(msg.sender, self, delta), "token-transfer-failed");
+        token.burn(self, redeemAmount);
+        return;
+    } 
+
+    uint delta = safeSub(currentRedeemAmount, redeemAmount);
+    if (delta > 0) {
+      token.mint(msg.sender, delta);
+      require(token.transferFrom(self, msg.sender, delta), "token-transfer-failed");
+    }
+}
+
+// the disburse function can be used after an epoch is over to receive currency and tokens
+function disburse(uint epochID) public {
+    // require epoch is settled
+    require((epochs[epochID].tokenPrice > 0), "epoch-not-settled-yet");
+        
+    uint currencyAmount = calcCurrencyDisbursement(epochID);
+    uint tokenAmount = calcTokenDisbursement(epochID);
+    
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = 0;
+    if (currencyAmount > 0) {
+        require(currency.transferFrom(self, msg.sender, currencyAmount), "currency-transfer-failed"); 
+    }
+    
+    epochs[epochID].redeemTokenAmount[msg.sender] = 0;
+    if (tokenAmount > 0) {
+        token.mint(msg.sender, tokenAmount);
```

Comment:
here the same. I think we will get in trouble with the tokenPrice, if we do it like that.